2021-year-report.org

```diff
@@ -0,0 +1,438 @@
+
+see: https://godbolt.org/z/PEGEWae4f
+
+**** Qualified Paths
+
+As part of my work into traits, this includes qualified paths. I have been investigating some test cases around qualified paths, and this one had me confused: https://github.com/rust-lang/rust/blob/master/src/test/ui/qualified/qualified-path-params-2.rs.
+
+The associated path <S as Tr>::A on its own will resolve to unit-struct S so when the final segment of ::f<u8> i would have assumed this would have resolved to the type of the impl function f substituted with u8. I don't see why this is ambiguous. I have, however, received an explanation on the Rust Zulip server https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Ambiguous.20associated.20types/near/251210283.
+
+This is interesting since associated types are not supported on impl blocks; the rustc compiler will enforce that all associated types must come from QualifiedPaths. This is to force the path to always be of the form <A as B>::C. When inherently associated types are stabilized, this will need some work. The GCC Rust implementation currently ICE's for this test case see https://github.com/Rust-GCC/gccrs/issues/843.
+
+See: [[https://doc.rust-lang.org/reference/paths.html#qualified-paths][QualifiedPathInType]]
+
+**** Never Type
+
+Rust has a never type which is pretty interesting. In c/c++, you can use the attribute no_return to signify that this function is never going to return. In Rust, they take this concept to its full with the never type, so for example, below: 
+
+#+BEGIN_SRC rust
+fn test() {
+   let a = return;
+   let b = a + 123;
+   a = 456;
+}
+#+END_SRC
+
+You can see 'a' is equal to the return expression, which will, in this context, never return. So even though everything after this, the first assignment will never get executed because of the return expression, Rust can fully type this function.  The functionality here is down to a key piece of the type system, which is more evident in closures than here. Still, in the above 'b = a + 123', the types here never type plus an integer inference variable, as there is no implementation of never type plus integer, here Rust actually returns another inference variable with a specified type-bound that requires the ability to add numbers together. This means the final assignment allows the type system to know that 'a' is an integer, albeit an uninitialized one, but this code is still safe because of the return expression.
+
+**** Method resolution 
+
+Method resolution is a multi-variant task; it involves the autoderef mechanism, but more importantly, there is an implicit ordering of the candidates that is key as plain old Impl block methods are always prefered over trait's. 
```

Comment:
The "implicit ordering" is only "documented in code" IIRC? Maybe that's worth mentioning here.